Initial round of revisions (#1)
alicederyn/peps: pep-9999.rst

        else:
            return f'{movie["name"]} ({movie["year"]})'

> A type checker should then consider this code type safe::

Author: I took this language straight from https://peps.python.org/pep-0589/ -- happy to change if you still prefer it

Reviewer: I see. As I mentioned, this is mostly "nit"; I think my wording is slightly more explicit but I'm fine either way

Take for example the function ``movie_string`` defined above. Suppose its type signature is defined using a ``TypedDict``::

    from typing import NotRequired, TypedDict

    class Movie(TypedDict):
        name: str
        year: NotRequired[int | None]

    class MovieRecord(TypedDict):
        name: str
        year: int

    def movie_string(movie: Movie) -> str:
        if movie.get("year") is None:
            return movie["name"]
        else:
            return f'{movie["name"]} ({movie["year"]})'

Attempting to pass a ``MovieRecord`` into ``movie_string`` results (using mypy) in the error::

    Argument 1 to "movie_string" has incompatible type "MovieRecord"; expected "Movie"

> This particular case case should be type-safe, but the type checker correctly stops the

Author: ```suggestion
This particular use case should be type-safe, but the type checker correctly stops the
```

        @overload
        def __getitem__(self, key: Literal["year"]) -> int | None: ...

This is very repetitive, easy to get wrong, and is still missing important method definitions like ``__contains__()`` and ``keys()``.

The proposed ``TypedMapping`` type allows a straightforward way of defining these types that should be familiar to existing users of ``TypedDict`` and support the cases exemplified above.

> In addition to those benefits, by flagging arguments of a function as ``TypedMapping`` makes explicit not just to typecheckers but also to user that the function is not going to modify its argument, which is usually a desirable property of a function interface.

Author: ```suggestion
In addition to those benefits, by flagging arguments of a function as ``TypedMapping``, it makes explicit not just to typecheckers but also to users that the function is not going to modify its inputs, which is usually a desirable property of a function interface.
```

Reviewer: 👍

> Finally, this allows bringing the benefits of ``TypedDict``to other mapping types that are unrelated to ``dict``.

Author: ```suggestion
Finally, this allows bringing the benefits of ``TypedDict`` to other mapping types that are unrelated to ``dict``.
```

Specification
=============

The ``TypedMapping`` type is a protocol behaving almost identically to ``TypedDict``, except:

> 1. Its instances need not be instances of subclasses of ``dict``

Author: ```suggestion
1. The runtime type is not constrained to be a ``dict``
```
Realised we can simplify this as, according to https://peps.python.org/pep-0589/:
> the runtime type of a TypedDict object will always be just dict (it is never a subclass of dict)

Reviewer: That's a simpler way to specify the constraint, I like it. Still I'd like the subject to be more explicit, e.g. "The runtime type of conforming values is not constrained to be a ``dict``"

Author: ```suggestion
1. The runtime type of a TypedMapping object is not constrained to be a ``dict``
```

    movie: MovieRecord = { "name": "Blade Runner",
                           "year": 1982 }
     
    movie["year"] = 1985  # Fine, mutator methods added in definition
    movie["name"] = "Terminator"  # Type check error, "name" mutator not declared

> Type Consistency Rules

Author: Copied this heading from https://peps.python.org/pep-0589/ too but happy to change it :)

Reviewer: Not my fault that mine is better :P (still, this is a "nit")
